apps/src/applab/AppLabView.jsx

```diff
@@ -37,23 +48,64 @@ var AppLabView = React.createClass({
           onScreenCreate={this.props.onScreenCreate} />;
     }
 
+    var instructionsHeight = this.props.instructionsHeight;
+    if (this.props.instructionsCollapsed) {
+      instructionsHeight = HEADER_HEIGHT;
+    }
+
+    if (!this.props.instructionsInTopPane) {
+      instructionsHeight = 0;
+    }
+
+    var codeWorkspaceContainerStyle = {
+      top: instructionsHeight
+    };
+
+
+    // TODO - there are a small set of levels that have instructions but not markdownInstructions
+    //   (that are also used in scripts). should convert these to have markdown instructions
+    // TODO - changing id of codeWorkspace to codeWorkspaceApplab will break callouts and some UI tests
+    // TODO - could group rightSide into single component?
     return (
       <ConnectedStudioAppWrapper>
         <div id="visualizationColumn">
           {playSpaceHeader}
           <ProtectedStatefulDiv contentFunction={this.props.generateVisualizationColumnHtml} />
         </div>
-        <ProtectedStatefulDiv id="visualizationResizeBar" className="fa fa-ellipsis-v" />
-        <ProtectedStatefulDiv id="codeWorkspace">
-          <ProtectedStatefulDiv id="codeWorkspaceWrapper" contentFunction={this.props.generateCodeWorkspaceHtml}/>
-          {!this.props.isReadOnlyWorkspace && <ProtectedStatefulDiv id="designWorkspace" style={{display: 'none'}} />}
-        </ProtectedStatefulDiv>
+        <ProtectedStatefulDiv
+            id="visualizationResizeBar"
+            className="fa fa-ellipsis-v" />
+        <TopInstructions
+            height={instructionsHeight}
+            markdown={this.props.instructionsMarkdown}
+            collapsed={this.props.instructionsCollapsed}
+            onToggleCollapsed={this.props.toggleInstructionsCollapsed}
+            onChangeHeight={this.props.setInstructionsHeight}/>
+        <CodeWorkspaceContainer
+            topMargin={instructionsHeight}
+            generateCodeWorkspaceHtml={this.props.generateCodeWorkspaceHtml}
+            onSizeChange={utils.fireResizeEvent}/>
       </ConnectedStudioAppWrapper>
     );
   }
 });
 module.exports = connect(function propsFromStore(state) {
   return {
-    isReadOnlyWorkspace: state.level.isReadOnlyWorkspace
+    isReadOnlyWorkspace: state.level.isReadOnlyWorkspace,
+    instructionsInTopPane: state.instructions.inTopPane,
+    instructionsMarkdown: state.level.instructionsMarkdown,
+    // TODO - how do i actually feel about using inTopPane here?
+    instructionsCollapsed: state.instructions.collapsed || !state.instructions.inTopPane,
+    instructionsHeight: state.instructions.height,
```

Comment:
Three different concepts affect how instructions are displayed: `instructionsInTopPane`, `instructionsCollapsed`, and `instructionsHeight`.

Right now `instructionsHeight` represents the height of instructions (including the header) if uncollapsed.

In our render function, `instructionsHeight` represents the height of the rendered instructions.

This means if collapsed, our `instructionsHeight` in render will be 30 and this.props.instructionsHeight will be larger. If not inTopPane, `instructionsHeight` in render will be 0 and `this.props.instructionsHeight` will not be. I wonder if there's a clearer way of accomplishing this.